modules/gateway/synchronize.go

```diff
@@ -34,26 +31,93 @@ func (g *Gateway) Synchronize() (err error) {
 // most recent block seen by both peers. From this starting height, it
 // transmits blocks sequentially. Multiple such transmissions may be required
 // to fully synchronize.
-func (g *Gateway) synchronize(peer network.Address) {
-	var newBlocks []consensus.Block
-	err := peer.RPC("SendBlocks", g.blockHistory(), &newBlocks)
-	if err != nil && err.Error() != moreBlocksErr.Error() {
-		// TODO: try a different peer?
+func (g *Gateway) synchronize(peer modules.NetAddress) {
+	for {
+		var newBlocks []consensus.Block
+		newBlocks, moreAvailable, err := g.requestBlocks(peer)
+		if err != nil {
+			// TODO: try a different peer?
+			return
+		}
+		for _, block := range newBlocks {
+			acceptErr := g.state.AcceptBlock(block)
+			if acceptErr != nil {
+				// TODO: If the error is a FutureTimestampErr, need to wait before trying the
+				// block again.
+			}
+		}
+
+		// loop until there are no more blocks available
+		if !moreAvailable {
+			break
+		}
+	}
+}
+
+// sendBlocks returns a sequential set of blocks based on the 32 input block
+// IDs. The most recent known ID is used as the starting point, and up to
+// 'MaxCatchUpBlocks' from that BlockHeight onwards are returned. It also
+// sends a boolean indicating whether more blocks are available.
+func (g *Gateway) sendBlocks(conn modules.NetConn) (err error) {
+	// read known blocks
+	var knownBlocks [32]consensus.BlockID
+	err = conn.ReadObject(&knownBlocks, 32*crypto.HashSize)
+	if err != nil {
 		return
 	}
-	for _, block := range newBlocks {
-		acceptErr := g.state.AcceptBlock(block)
-		if acceptErr != nil {
-			// TODO: If the error is a FutureTimestampErr, need to wait before trying the
-			// block again.
+
+	// Find the most recent block from knownBlocks that is in our current path.
+	found := false
+	var start consensus.BlockHeight
+	for _, id := range knownBlocks {
+		if height, exists := g.state.HeightOfBlock(id); exists {
```

Comment:
need to check that the block is also in the current path, you can use `BlockAtHeight`